velox/functions/prestosql/aggregates/ChecksumAggregate.cpp

```diff
@@ -0,0 +1,225 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/exec/Aggregate.h"
+#include "velox/functions/prestosql/aggregates/AggregateNames.h"
+#include "velox/functions/prestosql/aggregates/PrestoHasher.h"
+#include "velox/functions/prestosql/aggregates/SimpleNumericAggregate.h"
+
+namespace facebook::velox::aggregate {
+
+namespace {
+
+class ChecksumAggregate : public exec::Aggregate {
+ public:
+  static const long kPrime64 = 0x9E3779B185EBCA87L;
+
+  explicit ChecksumAggregate() : Aggregate(VARBINARY()) {}
+
+  void finalize(char** /* groups */, int32_t /* numGroups */) override {}
+
+  int32_t accumulatorFixedWidthSize() const override {
+    return sizeof(int64_t);
+  }
+
+  void initializeNewGroups(
+      char** groups,
+      folly::Range<const vector_size_t*> indices) override {
+    for (auto i : indices) {
+      *this->template value<int64_t>(groups[i]) = (int64_t)0;
+    }
+  }
+
+  void extractValues(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    auto* vector = (*result)->as<FlatVector<int64_t>>();
+    VELOX_CHECK(vector);
+    vector->resize(numGroups);
+
+    auto* rawValues = vector->mutableRawValues();
+    for (vector_size_t i = 0; i < numGroups; ++i) {
+      rawValues[i] = *value<int64_t>(groups[i]);
+    }
+  }
+
+  void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    VELOX_CHECK_EQ((*result)->encoding(), VectorEncoding::Simple::FLAT);
+    auto vector = (*result)->as<FlatVector<StringView>>();
+    VELOX_CHECK(
+        vector,
+        "Unexpected type of the result vector: {}",
+        (*result)->type()->toString());
+    vector->resize(numGroups);
+    uint64_t* rawNulls = getRawNulls(vector);
+
+    for (int32_t i = 0; i < numGroups; ++i) {
+      char* group = groups[i];
+      if (isNull(group)) {
+        vector->setNull(i, true);
+      } else {
+        clearNull(rawNulls, i);
+        int64_t val = *value<int64_t>(group);
+        auto buffer =
+            AlignedBuffer::allocate<char>(sizeof(int64_t), vector->pool());
+        memcpy(buffer->template asMutable<char>(), &val, sizeof(int64_t));
+        vector->set(i, StringView(buffer->as<char>(), sizeof(int64_t)));
+      }
+    }
+  }
+
+  void addRawInput(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /**/) override {
+    const auto& arg = args[0];
+
+    // TODO : Implement pushdown
+    // TODO : Clear null flag ?
+
+    DecodedVector decoded(*arg, rows, true);
+
+    if (!decoded.base()->type()->isPrimitiveType()) {
+      VELOX_NYI("Complex Types not yet supported!");
+    }
+
+    if (decoded.isConstantMapping()) {
+      auto isNull = decoded.isNullAt(0);
+
+      if (!isNull) {
+        rows.template applyToSelected(
+            [&](vector_size_t row) { computeHash(groups[row], decoded, 0); });
```

Comment:
Is `decoded` still accessed through `valueAt<T>(row)` at each row? It could be better if we can assign `valueAt<T>(0)` to a local variable before the `applyToSelected` loop and use this variable directly in the loop.

Reply:
Not sure i understood, but the value is accessed via getHash function , which takes decoded vector and then valueAt. The reason I am not assigning to a local variable is because that doesnt work for complex, nested types - as cost of materializing/copy can be high.

Reply:
@kgpai I think @kagamiori has a point. With the existing code, computeHash ends up calling getHash(decoded, 0) again an again for each row. Wouldn't it be better to calculate that value once, then add it to existing hashes in a loop.